kore/test/Test/Kore/Step/Axiom/Matcher.hs

```diff
@@ -1029,6 +1040,195 @@ test_unificationWithAppMatchOnTop =
+        assertEqualWithExplanation "" expect actual
+    , testCase "variable" $ do
+        let expect = substitution [(Mock.xBool, True)]
+        actual <- matchVariable Mock.xBool True
+        assertEqualWithExplanation "" expect actual
+    ]
+  where
+    top = Just $ OrPredicate.fromPredicate Predicate.topPredicate
+    substitution subst = Just $ MultiOr.make
+        [ Conditional
+            { term = ()
+            , predicate = makeTruePredicate
+            , substitution =
+                Substitution.unsafeWrap
+                    ((fmap . fmap) mkBool subst)
+            }
+        ]
+    mkBool = Bool.asInternal Mock.boolSort
+    matchConcrete = matchDefinition `on` mkBool
+    matchVariable var val =
+        matchDefinition (mkVar var) (mkBool val)
+
+matchingInt :: [TestTree]
+matchingInt =
+    [ testCase "concrete top" $ do
+        let expect = top
+        actual <- matchConcrete 1 1
+        assertEqualWithExplanation "" expect actual
+    , testCase "concrete bottom" $ do
+        let expect = Nothing
+        actual <- matchConcrete 1 0
+        assertEqualWithExplanation "" expect actual
+    , testCase "variable" $ do
+        let expect = substitution [(Mock.xInt, 1)]
+        actual <- matchVariable Mock.xInt 1
+        assertEqualWithExplanation "" expect actual
+    ]
+  where
+    top = Just $ OrPredicate.fromPredicate Predicate.topPredicate
+    substitution subst = Just $ MultiOr.make
+        [ Conditional
+            { term = ()
+            , predicate = makeTruePredicate
+            , substitution =
+                Substitution.unsafeWrap
+                    ((fmap . fmap) mkInt subst)
+            }
+        ]
+    mkInt = Int.asInternal Mock.intSort
+    matchConcrete = matchDefinition `on` mkInt
+    matchVariable var val =
+        matchDefinition (mkVar var) (mkInt val)
+
+matchingList :: [TestTree]
+matchingList =
+    [ testCase "concrete top" $ do
+        let expect = top
+        actual <- matchConcrete [1, 2] [1, 2]
+        assertEqualWithExplanation "" expect actual
+    , testCase "concrete bottom" $ do
+        let expect = Nothing
+        actual <- matchConcrete [1, 2] [1, 3]
+        assertEqualWithExplanation "" expect actual
+    , testCase "single variable" $ do
+        let expect = substitution [(Mock.xInt, 2)]
+        actual <- matchVariable [Right 1, Left Mock.xInt] [1, 2]
+        assertEqualWithExplanation "" expect actual
+    , testCase "two variables" $ do
+        let expect = substitution [(Mock.xInt, 2), (Mock.yInt, 4)]
+        actual <- matchVariable
+            [ Right 1
+            , Left Mock.xInt
+            , Right 3
+            , Left Mock.yInt
+            ]
+            [ 1, 2, 3, 4 ]
+        assertEqualWithExplanation "" expect actual
+    , testCase "no AC" $ do
+        let expect = Nothing
+        actual <- matchVariable
+            [ Right 1, Left Mock.xInt] [ 2, 1 ]
+        assertEqualWithExplanation "" expect actual
```

Comment:
Add `match [ 1 , 2 ] [ Left Mock.xInt , Left Mock.yInt ]`.

Also can we do `match Mock.xList [ 1 , 2 ]`, which should have substitution `Mock.xList -> [ 1 , 2 ]`?

And we should be able to match the tail of a list too, something like `match [ Mock.xInt , Mock.yList ] [ 1 , 2 , 3 ]` where `Mock.xInt -> 1` and `Mock.yList -> [ 2 , 3 ]`.